### ISSUE TITLE
Cache Maps as Web Resources

### ISSUE THREAD

**Issue by icing** [EricssonResearch/blind-cache-draft#9]
There are several questions raised in the draft related to the management (transferal, invalidation, update) of cache maps. One solution to these would treat cache maps as real web resources. The following examples try to describe the changes and benefits of such a proposal:

Section 5.2.1 (Initial Request): The description covers the case that the resource requested by the client is oob-cachable. However, for many sites the page, requested first, is dynamic  and will not be placed in a OOB-Cache. Instead, several static sub resources will. The client will only then learn about a possible cache map in 222 answers. When using HTTP/1.1 this will happen on several connections at the same time (on HTTP/2 in several streams on the same connection). This uses server and client resource unnecessarily.

Making the Cache Map a resource on the origin server (individual for the client or not), allows the server to sent the URL of the map already in a response header on the first request. Updating such a map is done by giving the updated map a new URL and sending that to the client on the next request. The client will see that the URL of the cache map has changed, that it does not reside in its local cache, and it will the updated map again. Also, normal HTTP cache-control mechanisms can be applied to the cache maps.

Example:
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-map"
```

Section 5.2.2: first bullet about cache miss. In this proposal, the client would sent a link header for the cache map used with a miss relation in the request. Such as
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-map-miss"
```
The origin server may then decide to send a new cache map link URL or answer the request directly with the resource content or do both.

Section 5.2.2: second bullet about cache unreachable. The link header on the request could be
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-unreachable"
```

Section 5.2.2: third bullet about the empty cache map. In this proposal, the origin server would indicate a new cache map, which has no mappings:
```
Link: <https://origin/maps/empty>;rel="oob-cache"
```
or one would define a specific link which identifies the NULL map.
```
Link: <emptyxxx>;rel="oob-cache"
```
where `emptyxxx` is a tbf const URL for the empty cache map.

Section 5.2.3: notifying the origin about compromised cache resources:
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-compromised"
```

Another standard attribute for links is the `anchor` which makes clear relative to what resource the uri rerefence of the link should be interpreted. It might be possible to denote the connection between
cache map and oob cache with it, as in:

Indication which cache the map is for:
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-map";anchor="http://bc.org/"
```

Indication which resource failed integrity check:
```
Link: <https://origin/maps/xx5hjsmd>;rel="bc-compromised";anchor="http://bc.org/3293232/cjkdadsk7893"
```

**Comment by icing** [comment]
A cache map might be a JSON resource like this:
```
{
   bc: "https://bc1.org",
   resources: [
     { from: "/js/jquery.js", to: "/dsjkds92/dsklds/dsds", headers: {
         "Last-Modified", "..."
     } },
     { from: "/img/logo.png", to: "/dsjkdds22/dsd2/d4ds", headers: {
         "Last-Modified", "..."
     } },
    ...
   ]
}
```

**Comment by gaperik** [comment]
I like the look of this.  Let's chat about this at the next hangout.
Drawbacks?

2015-09-11 14:19 GMT+02:00 Stefan Eissing <notifications@github.com>:

**Comment by gaperik** [comment]
I noted to mentioned you write a "resource on the origin server"- that is an important limitation security wise, right that the cache-map need to be delivered from the same origin? Should probably be made explicit. I also think one should be clear on what is mean by same origin and how this is asserted by the client.

**Comment by DanielLindstrm** [comment]
I think I like the idea.. Ill need to think more about it though. :-)
I guess there would be no problem combining this with h2push? I.e. pushing the map in the background and then it's there when the client needs it.. That should be possible to implement from service-worker also 

If you want to invalidate maps from the client I guess you can just point to a map and saying rel="oob-objects-removed" or you can put something in the map indicating the specific object was removed.

**Comment by icing** [comment]
Security: I'll write something in another issue. We need some more details and classifications since OOB caches will live in a variety of scenarios.

Re h2push: yes, cache maps would be an excellent use case for server pushes.

Invalidation: the client could indicate which resource had problems in the fragment of the link:

```
Link: <https://origin/maps/xx5hjsmd#3>;rel="bc-compromised"
```

Other ways are if course also possible. If the server then throw the map away, is a matter of server policy and implementation. Maybe the server does not trust clients.

**Comment by DanielLindstrm** [comment]
If just one object has been deleted, if you use
<pre>Link: <https://origin/maps/xx5hjsmd#3>;rel="bc-compromised"</pre>
the idea would be that only the map for that specific object is in that resource?

**Comment by icing** [comment]
I don't know if the protocol needs to specify that. The server could do that or delete the whole map or remove that whole oob cache from its list. Or do nothing and wait for a second client to report a problem, or try first a retrieval itself...

**Comment by DanielLindstrm** [comment]
I think that it would probably be good to specify how to update maps and also how to remove objects from the map.. But anyway.. I don't see any problems doing that.. it's more about selecting how to.

Just to check so we are thinking the same.. I see that the map linked to does not need to contain the complete set of objects, but it can just contain a subset and the map can later be extended with more objects via new Link headers.
I also see that a map can be "personal" for a specific user in case some of the objects is only intended for only that user or for a group of users.

**Comment by icing** [comment]
Correct. And the server can keep a number of maps, if it wants. It's a balance between map size and number of maps the clients needs to cache. Making the map too large will mean unnecessary transfer, making it too small will clutter the client cache.

Adding a resource might add a map in order to avoid invalidating an existing one. A video service might keep a separate map for each movie it serves. Another server might have a single map for each authenticated user.

**Comment by gaperik** [comment]
Meaning it is easy to devOp and beta test a new server version for handling
the cache-map and gradually add it. This is the strong part of this.

But what about cache misses- should the error be reported to the origin or
to the cache map server (module in the same origin)? Or should the
cache-map server get it or both?

2015-09-15 13:22 GMT+02:00 Stefan Eissing <notifications@github.com>:

**Comment by DanielLindstrm** [comment]
Maybe the bc address could be the "key" to allow multiple bc in a map?
Like this..
```
{
   "https://bc1.org": {
     map: [
       { from: "/js/jquery.js", to: "/dsjkds92/dsklds/dsds", headers: {
           "Last-Modified", "..."
       } },
       { from: "/img/logo.png", to: "/dsjkdds22/dsd2/d4ds", headers: {
           "Last-Modified", "..."
       } },
      ...
     ]
    },
 "https://bc2.org": {
     map: [
       { from: "/img/abc.jpg", to: "/dsj321/21c3/223f", headers: {
           "Last-Modified", "..."
       } },
       { from: "/img/car.jpg", to: "/dsj321/21c3/f33f", headers: {
           "Last-Modified", "..."
       } },
      ...
     ]
    }
  }
  ```

**Comment by gaperik** [comment]
Ok. I propose we outline a draft text for the cache-map as a resource.
